### ISSUE TITLE
(🐞) false narrowing with reassignment in loop

### ISSUE THREAD

**Issue by KotlinIsland** [python/mypy#11612]
```py
a: int | None
a = None
for i in range(10):
    reveal_type(a)  # "None" and also "int | None"
    if a:
        print(a)  # error: unreachable
    a = i
```

Also:
```py
a = None
for b in [1, 2]:
    if a is None:
        a = b
        continue
    print("hi")  # false unreachable error
```

**Comment by sobolevn** [comment]
This can be fixed by:

```python
a: int | None = None
for b in [1, 2]:
    if a is None:
        a = b
        continue
    print("hi")  # ok
```

The other question is: why `a` is not infered to be `None | int`?

**Comment by KotlinIsland** [comment]
It is

**Comment by KotlinIsland** [comment]
@sobolevn It's not just the inference, I've updated the OP with a much better example.
